src/java/test/org/apache/zookeeper/server/ZooKeeperServerMainTest.java

```diff
@@ -138,14 +145,56 @@ void delete(File f) throws IOException {
         ServerCnxnFactory getCnxnFactory() {
             return main.getCnxnFactory();
         }
+
     }
 
-    public static  class TestZKSMain extends ZooKeeperServerMain {
+    public static class TestZKSMain extends ZooKeeperServerMain {
+
+        private ServerStats serverStats;
+
+        @Override
+        public ZooKeeperServer getZooKeeperServer(FileTxnSnapLog txnLog, ServerConfig config, ZKDatabase zkDb) {
+            ZooKeeperServer zooKeeperServer = super.getZooKeeperServer(txnLog, config, zkDb);
+            serverStats = zooKeeperServer.serverStats();
+            return zooKeeperServer;
+        }
+
+        @Override
         public void shutdown() {
             super.shutdown();
         }
     }
 
+    // Test for ZOOKEEPER-2770 ZooKeeper slow operation log
+    @Test
+    public void testRequestWarningThreshold() throws IOException, KeeperException, InterruptedException {
+        ClientBase.setupTestEnv();
+
+        final int CLIENT_PORT = PortAssignment.unique();
+
+        MainThread main = new MainThread(CLIENT_PORT, true, null, 0);
+        main.start();
+
+        Assert.assertTrue("waiting for server being up",
+                ClientBase.waitForServerUp("127.0.0.1:" + CLIENT_PORT,
+                        CONNECTION_TIMEOUT));
+        // Get the stats object from the ZooKeeperServer to keep track of high latency requests.
+        ServerStats stats = main.main.serverStats;
+
+        ZooKeeper zk = new ZooKeeper("127.0.0.1:" + CLIENT_PORT,
+                ClientBase.CONNECTION_TIMEOUT, this);
+
+        zk.create("/foo1", "foobar".getBytes(), Ids.OPEN_ACL_UNSAFE,
+                CreateMode.PERSISTENT);
+
+        Assert.assertEquals(new String(zk.getData("/foo1", null, null)), "foobar");
+        // It takes a while for the counter to get updated sometimes, this is added to reduce flakyness
+        Thread.sleep(1000);
```

Comment:
OK. The point of the timeout, however, is that an update has not been applied. The sync() is designed to wait for exactly this.

Code that depends on timeouts is not deterministic in my book. It may just have a low probability of failure.

If this isn't feasible, however, it may be something that has to be lived with.